server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java

```diff
@@ -1817,35 +1817,27 @@ public Metadata build(boolean builtIndicesLookupEagerly) {
                     assert dataStream.getIndices().isEmpty() == false;
 
                     List<String> aliases = dataStreamToAliasLookup.getOrDefault(dataStream.getName(), List.of());
+                    final IndexAbstraction.DataStream dsAbstraction = new IndexAbstraction.DataStream(dataStream, aliases);
                     IndexAbstraction existing = indicesLookup.put(
                         dataStream.getName(),
                         new IndexAbstraction.DataStream(dataStream, aliases)
                     );
                     assert existing == null : "duplicate data stream for " + dataStream.getName();
 
                     for (Index i : dataStream.getIndices()) {
-                        indexToDataStreamLookup.put(i.getName(), dataStream);
+                        indexToDataStreamLookup.put(i.getName(), dsAbstraction);
```

Comment:
I wonder if we need to build `indextToDataStreamLookup` at all? Could we not simply populate `indicesLookup` directly here instead and ignore those below (asserting that they are data stream owned instead of the duplicate check)?

Reply:
Right we don't ... now comes the interesting question of whether this would actually be faster. 

Looking at the profiling for the full method (after my change here):

<img width="2259" alt="image" src="https://user-images.githubusercontent.com/6490959/151545875-2e242283-b744-4275-bfbf-ad05bf769257.png">

for a completely degenerate case where we have about as many indices as datastreams (one index per DS) ...
the hash map building barely shows up relative to the tree-map. 

So the question becomes, is it cheaper to do 2 puts (or I guess one put and one computeIfAbsent) to the tree-map (the second put in the indices loop would be to a larger tree map than the one in the DS loop) or do one put into an on average smaller tree map and instead build the technically redundant hash-map.

Profiling suggests that this is faster I'd say. And in the real world, where you'd have multiple indices per DS the difference would be even more visible.
The problem we're running into is that the tree-map is super lob sided because we have the shared `.ds` prefix for all indices pretty much and then some more sharing for the specific DS, leading to loads of comparisons.
We should've used a different naming schema for the internal DS indices I guess ... (in fact I think we still could).